### ISSUE TITLE
visualisation of embedding space

### ISSUE THREAD

**Issue by ghost** [hrbigelow/ae-wavenet#3]
Hi, nice work you've done! I was wondering if you had any visualisation of trained vq embedding space? I've been trying to reproduce the paper as well, but the model seems not to use the global condition. It seems the wavenet is too powerful to pick up local / global condition, and only uses the original wav audio. Is that what you are facing with the codebook collapse issue as well?

**Comment by hrbigelow** [comment]
Hi Dongyao,

Thanks for your interest.  So far, I haven't plotted the embedding space.  So, as to your first question, they do use global condition, the speaker ID, by concatenating it to the 128D vector (see the "concat" box in Figure 1).  I've been using a 10D vector for the global embedding.

To answer your second question, yes, that does seem to be a hazard of this setup, and they mention it in section C, third paragraph: "Training mixed latent variable and autoregressive models
is prone to latent space collapse...", however, they claim this problem is solved by the quantization.  I've had personal communication with others though who think that EMA is necessary as well.

So, I don't know whether this is exactly what I'm seeing.  One thing I saw, after ~1000 steps or so, was that the encoder actually learns to ignore the input (the activations of the third layer go to zero), and because the remaining layers all have residual and bias, is able to output non-zero latents.  The decoder's gradients w.r.t these latents slowly diminish as the decoder learns they are not helpful.

But, that issue is mitigated a bit by lowering the learning rate or increasing batch size.  Unfortunately, I have only Google Colab to work with.  That gives me about 6000 training steps every 12 hours, and they trained for 800,000 steps!

Do you have access to more powerful compute?  Currently, I'm trying to get it working on TPU using torch_xla.

**Comment by ghost** [comment]
Hey,
Thanks for the reply.

I did use a speaker embedding in the model, what I meant was that the model did learn to separate speaker id through speaker embedding, but didn't seem to make use of it in generating. All the generated audio are like the same id.

I've tried ema, but it didn't give better results than without it. I'm actually trying without the additional stuff in the 2019 paper, but just the original structure from 2016.
For the encoder part, what I've been seeing is that both the mean and variance of the vq embedding space converges to around [-0.05, 0.05] within like 30k steps. I didn't check the gradients though.

I have access to a 2080Ti, but can only use batch size of up to 4, since I'm using 3 x 10 layers of kernel size 3 and 512 filters in wavenet. I could get to batch size of 8 with less filters, but I'm not sure if batch size matters a lot here. I had a learning rate schedule starting from 0.001.
BTW, would you mind going to my repo and take a look at the tensorboard visualisation, and see if any of the graphs means anything to you?

**Comment by hrbigelow** [comment]
Hi,

When you say the model "learned to separate speaker id through speaker embedding", what do you mean?  In the Chorowski paper, the encoder has no knowledge of speaker ID, and in the wavenet decoder, the speaker ID is already input as separate one-hot channels catenated with the encoder output vectors.  So, doesn't that mean it is already "separated"?  Or are you saying something different?

Also, for the latent-space collapse issue, you might be interested in https://arxiv.org/abs/1906.00446 (VQ-VAE2) because they actually now move the entire autoregressive part into the model for the prior.  And, so, with this new design, the decoder must now rely entirely on the latent vectors - it has no chance of ignoring anything.  I didn't see any comment in that paper about the decision to do that, but I think it makes sense.

One other problem that I had was something not mentioned in the paper.  That is, I found that the number of different codebook vectors (if I started with 1000, say) that were "active" (recently used as nearest neighbors for some encoder output) became more and more concentrated to a few vectors.  If I use their L2 norm loss terms for both the nearest neighbor calculation and commitment / L2 loss, there is an unfortunate boundary phenomenon in which the codebook vector that happens to have the largest norm will easily be the nearest neighbor if the encoder tends to produce longer vectors than the codebook.  This boundary phenomenon wouldn't happen if the geometry were different.  So, currently I'm experimenting with an extra abs(||c|| - 1) loss, which pushes the codebook vectors towards the unit sphere.  In that geometry, I hope, there won't be any special vectors that are on the boundary of the space.  Not sure if it will work or not.

I would be happy to try out your repo.  Will let you know.

**Comment by ghost** [comment]
Hey,

For the speaker / global condition, instead of an one-hot vector, I used a trainable embedding space (so it could be better compared before / after training). After training, the embedded vectors are separated between genders. Although this doesn't tell anything about whether or not the wavenet decoder discards speaker information from the raw audio and only uses from gc.

For the number of different codebook vectors being used (the q(z|x) I think?), my result shows that it did move to a uniform distribution, i.e. most of them are "active", but at first only some of them are used. Despite that, the mean and variance of the codebook space is so close to 0, that I assume they contribute little to the wavenet decoder as local condition. My experience from my many attempts is that the initialisation matters more than the regularisation in preventing collapse; for me a truncated normal init worked the best, but I assume it depends on how the encoder output is distributed originally.

I haven't read VQ-VAE2 yet... but I'll try it out.
